### PR TITLE
HED+PKG unpacker for the PC release of Kingdom Hearts

### PR DESCRIPTION
**Pull request by Xeeynamo**
Work in progress extractor. At the moment it is able to parse the HED table, decrypt and then decompress the files in the PKG. I tested it with `kh2_*.hed` and `SettingMenu.hed` and it works. So I presume it will also work with all the other games.

At the moment it is not able to extract the DDS textures that contains the HD assets. The biggest flaw for now is that it is not possible to obtain the full list of files. Infact the HED table contains a MD5 of the file name, not the file name itself. Luckily it seems that the HED table was created by storing the files in alphabetical order, so it might be easy to figure the original names. To research a file name, search for an online MD5 calculator, try your string and see if the MD5 matches.

I plan to release it when we are going to have a very good amount of file names extracted, which is a bit far from now. I need your help to figure out the file names.

* [X] Add decryption
* [X] Add decompression
* [X] Recognise if a file contains HD assets
* [X] Extract base file
* [x] Extract HD assets
* [X] All names for KH1
* [X] All names for KH2
* [X] All names for BBS
* [X] All names for KH3D
* [X] All names for ReCoM
* [x] All names for Mare
* [X] All names for SettingMenu
* [X] All names for Theater
* [X] All names for Launcher28

### Mare (2)

C193B7F53CF09EFF1CE33ACB48748D97 (.tm2)
FDF66B6D52944EFC92E9D639E9056EC1 (.tm2)

### CODE REVIEW

**Comment by Rikux3** [comment]
[mare_files.txt](https://github.com/Xeeynamo/OpenKh/files/6251631/mare_files.txt)
5 nonames:
C193B7F53CF09EFF1CE33ACB48748D97, FDF66B6D52944EFC92E9D639E9056EC1 => TM2
0BAAD3DE414ED65672D0E28C369794EE, 99866AFB76634A92939D7B84CE438782, D83F679B462F7C2F32D3EF8F93D190C3 => probably PNG, has `IEND` at the end of the file

**Comment by Noxalus** [comment]
For KH1FM, the unpacker currently found 3391 filenames and 368 are missing. Looking to the context, I found these two:

```
3E1638A06AF29B761E109A1D2111AEA0 => pi32.ard
33548D9E8C8E1D524AA4D96E40968A6C => pi37.ard
```

[founds.txt](https://github.com/Xeeynamo/OpenKh/files/6253761/founds.txt)
[missings.txt](https://github.com/Xeeynamo/OpenKh/files/6253763/missings.txt)

**Comment by efonte** [comment]
In this zip are all the hashes of the .hed files in the `Kingdom Hearts HD 1.5 +2.5 ReMIX / Image` folder.
There are 138 unknown hashes of 90984.
[hashes.zip](https://github.com/Xeeynamo/OpenKh/files/6253790/hashes.zip)

**Comment by Noxalus** [comment]
> In this zip are all the hashes of the .hed files from the `Kingdom Hearts HD 1.5+2.5 ReMIX/Image` folder.
> There are 138 unknown hashes of 90984.
> [hashes.zip](https://github.com/Xeeynamo/OpenKh/files/6253790/hashes.zip)

Thank you a lot, it completes all missing filenames for KH1FM except for the first hash of `kh1_first.hed`: `1C8F9652717EA8CA679417CE98807E92`.

It's a PNG file (so the name ends with `.png`) and it seems to be related to the "save data" (that what the image says). And the name should start with a "K" or a letter before it in the alphabet.

**Comment by Xeeynamo** [comment]
Marking the PR as ready to be reviewed.

File: OpenKh.Command.IdxImg/EpicGamesAssets.cs [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+                private static void CreateDirectoryForFile(string fileName)
+                {
+                    var directoryName = Path.GetDirectoryName(fileName);
+                    if (!Directory.Exists(directoryName))
+                        Directory.CreateDirectory(directoryName);
+                }
+            }
+
+            //private class ListCommand
```
Comment (Delta-473) [inline OpenKh.Command.IdxImg/EpicGamesAssets.cs:149]: Leftover copy from the KH2 command?

Comment (Xeeynamo) [reply OpenKh.Command.IdxImg/EpicGamesAssets.cs:149]: Right! Lemme fix that once I am back home

**Comment by efonte** [comment]
more hashes
```ini
cd9dc2c4486fa21540d7e91ecd384c42 = KHReCoM.png
fa897efca73d1a7e2e801318b71ed91d = menu/de/load_menu.l2d
2d687293a81fc0175ee6c533bf249522 = menu/en/load_menu.l2d
c765c97c056e4efe877a80ed47992850 = menu/es/load_menu.l2d
009a2c0eebeed7aea45e3174195c63ee = menu/fr/load_menu.l2d
ed8195c126962dc1ea6c292143ef41c5 = menu/it/load_menu.l2d
```

**Comment by ligeromiguel** [comment]
Here are the missing ones for the Setting Menu! :)

3973831EDA66382A9D7BC678136E5612 = fr/l2d/initialsetting.l2d
D1FC0D4E063B8807154DE81E430399E1 = jp/l2d/initialsetting.l2d
BAAA7D8408F3C77DA4C8D3F1639455D6 = gr/l2d/initialsetting.l2d
836FAD7D8C7E045B202BE9B953FB25D4 = sp/l2d/initialsetting.l2d
3BBC82C538D81D32BF2F9B330EFF4BBC = it/l2d/initialsetting.l2d
082816C3F444D3589DF15A8F81643E34 = uk/l2d/initialsetting.l2d

**Comment by Sirius902** [comment]
0BAAD3DE414ED65672D0E28C369794EE = KH15+25.png

**Comment by ligeromiguel** [comment]
> Here's a hash from Mare:
> 
> 0BAAD3DE414ED65672D0E28C369794EE = KH15+25.png

Thanks for your contribution, looks like the png from the KH2.8 Launcher is this following your discovery:

37FBA62E6E0EF7ECFC9D04F1526DA79C = KH28.png

**Comment by Sirius902** [comment]
Hey I managed to figure out the Theater one by guessing! That's all the `.png` files down.

49AE6D86B4395CD61A59A1B71E27E04F = KHFMTheater.png

**Comment by jthecodebuilder** [comment]
Based on the contents of the files, I've managed to find the last two filenames (+ paths) and complete the list! 😁

`C193B7F53CF09EFF1CE33ACB48748D97 = gr/tim2/theater/days_theatre.tm2
FDF66B6D52944EFC92E9D639E9056EC1 = gr/tim2/theater/Recoded_theatre.tm2`

**Comment by Truthkey** [comment]
> Based on the contents of the files, I've managed to find the last two filenames (+ paths) and complete the list! 😁
> 
> ```
> C193B7F53CF09EFF1CE33ACB48748D97 = gr/tim2/theater/days_theatre.tm2
> FDF66B6D52944EFC92E9D639E9056EC1 = gr/tim2/theater/Recoded_theatre.tm2
> ```

Thank you!

**Review by Vladabdf — DISMISSED** [review]
It looks like this is ready to be merged. I'll leave it open for a little bit longer just in case, but if another collaborator thinks it's good to go, it can be merged.

**Comment by Delta-473** [comment]
Doesn't #485 need to merged into this first?

**Comment by Vladabdf** [comment]
> 
> 
> Doesn't #485 need to merged into this first?

Oh, you're right. My mistake!

**Comment by Delta-473** [comment]
LGTM but didn't test it personally

**Comment by gledson999** [comment]
I've tested and I can confirm, this work in KH2 and BBS!
Others KH Games, I don't tested yet!

**Review by Vladabdf — APPROVED** [review]
Tested and confirmed working fully by a user by the name of EmreX_Coded.

**Pull request merged**